src/utils/contribution.js

```diff
@@ -60,51 +85,228 @@ export default class Contribution {
       ? then_
       : else_
   }
+
   /**
    * Return a specific class name if the condition of difference it true
    * @param  {} component original component
-   * @param  {} changes object containing changes
+   * @param  {} preview object containing changes
    * @param  {} field field to check
    * @param  {} className className to apply if the field has some changes
    */
-  static classIfDifferent(component, changes, field, className) {
-    return this.ifDifferent(component, changes, field, className, '')
+  static classIfDifferent(component, preview, field, className) {
+    return this.ifDifferent(component, preview, field, className, '')
   }
+
+  /**
+   * Check if the preview has some changes from the original definitions
+   * Returns an object containing each change
+   * @param {*} definition
+   * @param {*} preview
+   * @return {Object} Return a new object which represents the diff
+   */
+  static getChangesFromPreview(definition, preview) {
+    if (isEmpty(definition) || isEmpty(preview)) return
+    return this.difference(preview, definition)
+  }
+
+  /**
+   * Deep diff between two objects, using lodash
+   * @param  {Object} object Object compared
+   * @param  {Object} base   Object to compare with
+   * @return {Object} Return a new object which represents the diff
+   */
+  static difference(object, base) {
+    return transform(object, (result, value, key) => {
+      if (isArray(value)) {
+        return (result[key] = this.differenceBetweenObjects(object[key], base[key]))
+      }
+      if (!isEqual(value, base[key])) {
+        return (result[key] =
+          isArray(value) && isArray(base[key])
+            ? this.differenceBetweenObjects(result[key], base[key])
+            : isObject(value) && isObject(base[key]) && value)
+      }
+    })
+  }
+
+  // Compare 2 collections and returns a new array keeping the original keys
+  static differenceBetweenObjects(a, b) {
+    let difference = []
+    a.map((itemA, index) => {
+      let find = false
+      b.map(itemB => {
+        if (isEqual(itemA, itemB)) {
+          find = true
+          return false
+        }
+        return false
+      })
+      if (!find) {
+        difference[index] = itemA
+      }
+      return false
+    })
+    return difference
+  }
+
+  static printCoordinates = value => (value && value.url ? `${value.url}/commit/${value.revision}` : value)
+
+  static printDate = value => (!value ? null : moment(value).format('YYYY-MM-DD'))
+
+  static printArray = value => (value && isArray(value) ? value.join(', ') : null)
+
+  static getPercentage = (count, total) => Math.round(((count || 0) / total) * 100)
+
+  static isSourceComponent = component => ['github', 'sourcearchive'].includes(component.provider)
+
+  /**
+   *  Get image of definition based on the provider
+   *
+   * @param {*} item
+   * @returns image file
+   */
+  static getImage(item) {
```

Comment:
perhaps this should just be a table lookup?  providerImages?

Reply:
Thanks for your comment, @jeffmcaffer. May I ask you to clarify which table would we be looking up? `providerImages` might be a new property in the schema or is something already available? 

Probably we are just a little bit unclear on this particular comment and it might be of great help if you could add additional details?


by the way, `getImage` is a very simple switch so maybe we can postpone this particular improvement at a later stage?

Reply:
I'm thinking of a new table in contribution.js (or elsewhere). It is more to get rid of the big (and ever growing) switch statement. This can be deferred if you like